src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs

```diff
@@ -2,45 +2,138 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
 
 #if NETCOREAPP
 
-using System.Diagnostics;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
 using System.Reflection;
 using System.Runtime.Loader;
 
 namespace Microsoft.CodeAnalysis
 {
     internal class DefaultAnalyzerAssemblyLoader : AnalyzerAssemblyLoader
     {
-        private AssemblyLoadContext _loadContext;
+        internal static readonly ImmutableArray<string> CompilerAssemblySimpleNames =
+            ImmutableArray.Create(
+                "Microsoft.CodeAnalysis",
+                "Microsoft.CodeAnalysis.CSharp",
+                "Microsoft.CodeAnalysis.VisualBasic",
+                "System.Collections",
+                "System.Collections.Concurrent",
+                "System.Collections.Immutable",
+                "System.Console",
+                "System.Diagnostics.Debug",
+                "System.Diagnostics.StackTrace",
+                "System.IO.Compression",
+                "System.IO.FileSystem",
+                "System.Linq",
+                "System.Linq.Expressions",
+                "System.Memory",
+                "System.Reflection.Metadata",
+                "System.Reflection.Primitives",
+                "System.Resources.ResourceManager",
+                "System.Runtime",
+                "System.Runtime.CompilerServices.Unsafe",
+                "System.Runtime.Extensions",
+                "System.Runtime.InteropServices",
+                "System.Runtime.Loader",
+                "System.Runtime.Numerics",
+                "System.Runtime.Serialization.Primitives",
+                "System.Security.Cryptography.Algorithms",
+                "System.Security.Cryptography.Primitives",
+                "System.Text.Encoding.CodePages",
+                "System.Text.Encoding.Extensions",
+                "System.Text.RegularExpressions",
+                "System.Threading",
+                "System.Threading.Tasks",
+                "System.Threading.Tasks.Parallel",
+                "System.Threading.Thread",
+                "System.Threading.ThreadPool",
+                "System.Xml.ReaderWriter",
+                "System.Xml.XDocument",
+                "System.Xml.XPath.XDocument");
+
+        private readonly object _guard = new object();
+        private readonly Dictionary<string, AssemblyLoadContext> _loadContextByDirectory = new Dictionary<string, AssemblyLoadContext>();
```

Comment:
Please use explicit comparers when dealing with file paths as an indication you deliberately made the choice for this to be case sensitive.

```suggestion
        private readonly Dictionary<string, AssemblyLoadContext> _loadContextByDirectory = new Dictionary<string, AssemblyLoadContext>(StringComparer.Ordinal);
```